presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergEqualityDeleteAsJoin.java

```diff
@@ -0,0 +1,458 @@
+                        .build();
+
+                // ON source.delete_column = deletes.delete_column, ...
+                Set<ConnectorJoinNode.EquiJoinClause> clauses = deleteColumnAssignments
+                        .entrySet()
+                        .stream()
+                        .filter(f -> !IcebergMetadataColumn.isMetadataColumnId(((IcebergColumnHandle) (f.getValue())).getId()))
+                        .map(assignment -> {
+                            VariableReferenceExpression left = reverseAssignmentsMap.get(((IcebergColumnHandle) (assignment.getValue())).getId());
+                            VariableReferenceExpression right = assignment.getKey();
+                            return new ConnectorJoinNode.EquiJoinClause(left, right);
+                        }).collect(Collectors.toSet());
+
+                FunctionHandle lessThan = functionResolution.comparisonFunction(OperatorType.LESS_THAN, BigintType.BIGINT, BigintType.BIGINT);
+
+                // AND source.$data_sequence_number < deletes.$data_sequence_number
+                RowExpression versionFilter = new CallExpression(lessThan.getName(),
+                        lessThan,
+                        BooleanType.BOOLEAN,
+                        Collections.unmodifiableList(Arrays.asList(reverseAssignmentsMap.get(DATA_SEQUENCE_NUMBER.getId()), joinSequenceNumber)));
+
+                TableScanNode deleteTableScan = createDeletesTableScan(node,
+                        deleteColumnAssignments,
+                        icebergTableHandle,
+                        tableName,
+                        deleteFields,
+                        table,
+                        deleteGroupInfo);
+
+                parentNode = new ConnectorJoinNode(idAllocator.getNextId(),
+                        Arrays.asList(parentNode, deleteTableScan),
+                        Optional.empty(),
+                        ConnectorJoinNode.Type.LEFT,
+                        clauses,
+                        Sets.newHashSet(versionFilter),
+                        Stream.concat(parentNode.getOutputVariables().stream(), deleteTableScan.getOutputVariables().stream()).collect(Collectors.toList()));
+            }
+
+            FilterNode filter = new FilterNode(Optional.empty(), idAllocator.getNextId(), Optional.empty(), parentNode,
+                    new SpecialFormExpression(SpecialFormExpression.Form.IS_NULL, BooleanType.BOOLEAN,
+                            new SpecialFormExpression(SpecialFormExpression.Form.COALESCE, BigintType.BIGINT, deleteVersionColumns)));
+
+            return filter;
+        }
+
+        private static ImmutableMap<Set<Integer>, DeleteSetInfo> collectDeleteInformation(Table icebergTable,
+                IcebergTableHandle icebergTableHandle,
+                long snapshotId)
+        {
+            HashMap<Set<Integer>, DeleteSetInfo> deleteInformations = new HashMap<>();
```

Comment:
Use an Immutable Map builder

Reply:
This particular map can have duplicates added to it since there can be delete files with the same delete schema.  which fails the ImmutableMapBuilder. However, it is converted to an immutable map before being returned from this function.

Reply:
No problem, please add little comment explaining this